### PR TITLE
Fix null-deref thanks to an attribute on a global declarator chunk

### PR DESCRIPTION
**Pull request by erichkeane**
This was reported (sort of) in a PR: #77703. The problem is that a declarator 'owns' an attributes allocation via an `AttributePool`. However, this example tries to copy a DeclaratorChunk from one Declarator to another, so when the temporary Declarator goes out of scope, it deletes the attribute it has tried to pass on via the chunk.

This patch ensures that we copy the 'ownership' of the attribute correctly, and adds an assert to catch any other casess where this happens.

Additionally, this was put in as a bug report, so this
Fixes #83611

### CODE REVIEW

**Review by shafik — COMMENTED** [review]
Note, I opened an issue for this here: https://github.com/llvm/llvm-project/issues/83385

File: clang/lib/Sema/ParsedAttr.cpp [hunk 1]
```diff
@@ -100,6 +100,12 @@ void AttributePool::takePool(AttributePool &pool) {
   pool.Attrs.clear();
 }
 
+void AttributePool::takeFrom(ParsedAttributesView &List, AttributePool &Pool) {
+  assert(&Pool != this && "AttributePool can't take attributes from itself");
+  llvm::for_each(List.AttrList, [&Pool](ParsedAttr *A) { Pool.remove(A); });
```
Comment (shafik) [inline clang/lib/Sema/ParsedAttr.cpp:105]: Should we remove from the pool after we added them to the `Attrs`?

Comment (erichkeane) [reply clang/lib/Sema/ParsedAttr.cpp:105]: You mean reverse the order of the remove/add?  Based on `takeAllFrom` and `takeOneFrom` (https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Sema/ParsedAttr.h#L953), we are consistently inconsistent :)

In this case, I was working from `takeOneFrom` so I ended up in this order, but if there is a good reason to prefer one over the other, I can switch.

**Comment by erichkeane** [comment]
Note to self when I get back to work:
1- Update commit message with bug #[ 83611](https://github.com/llvm/llvm-project/pull/83611)
2- Add bug # to release note.

**Comment by erichkeane** [comment]
> Note to self when I get back to work:

Both done now!

File: clang/include/clang/Sema/ParsedAttr.h [hunk 2]
```diff
@@ -734,6 +735,9 @@ class AttributePool {
     pool.Attrs.clear();
   }
 
+  /// Take a list of attributes from another pool and add them to this pool.
```
Comment (AaronBallman) [inline clang/include/clang/Sema/ParsedAttr.h:738]:
```suggestion
  /// Removes the attributes from \c List, which are owned by \c Pool, and adds them at the end
  /// of this \c AttributePool.
```

**Review by AaronBallman — APPROVED** [review]
LGTM aside from a tiny commenting nit (take it or leave it).

**Review by shafik — APPROVED** [review]

**Pull request merged**